### PR TITLE
AF-555: [Preferences] Allow preferences properties to have help text, validation and form options.

### PR DESCRIPTION
**Pull request by paulovmr**

### CODE REVIEW

**Comment by paulovmr** [comment]
Part of an ensemble:
* https://github.com/AppFormer/uberfire/pull/754
* https://github.com/kiegroup/guvnor/pull/466
* https://github.com/kiegroup/kie-wb-common/pull/944

File: uberfire-extensions/uberfire-wires/uberfire-wires-webapp/src/main/java/org/uberfire/ext/wires/shared/preferences/bean/MyPreference.java [hunk 1]
```diff
@@ -16,34 +16,44 @@
...
 import org.uberfire.preferences.shared.annotations.WorkbenchPreference;
 import org.uberfire.preferences.shared.bean.BasePreference;
+import org.uberfire.preferences.shared.impl.validation.NotEmptyValidator;
 
 @WorkbenchPreference(identifier = "MyPreference",
         bundleKey = "MyPreference.Label")
 public class MyPreference implements BasePreference<MyPreference> {
 
-    @Property(bundleKey = "MyPreference.Text")
+    @Property(bundleKey = "MyPreference.Text",
```
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-webapp/src/main/java/org/uberfire/ext/wires/shared/preferences/bean/MyPreference.java]: Formatting in this class seems wrong.

File: uberfire-preferences/uberfire-preferences-api/src/test/java/org/uberfire/preferences/shared/impl/validation/NotEmptyValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package org.uberfire.preferences.shared.impl.validation;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class NotEmptyValidatorTest {
```
Comment (manstis) [inline uberfire-preferences/uberfire-preferences-api/src/test/java/org/uberfire/preferences/shared/impl/validation/NotEmptyValidatorTest.java]: Include test for non-```String``` values?

File: uberfire-preferences/uberfire-preferences-backend/src/test/java/org/uberfire/preferences/backend/MyPreference.java [hunk 3]
```diff
@@ -16,16 +16,21 @@
...
 import org.uberfire.preferences.shared.annotations.WorkbenchPreference;
 import org.uberfire.preferences.shared.bean.BasePreference;
+import org.uberfire.preferences.shared.impl.validation.NotEmptyValidator;
 
 @WorkbenchPreference(identifier = "MyPreference",
         bundleKey = "MyPreference.Label")
 public class MyPreference implements BasePreference<MyPreference> {
 
-    @Property(bundleKey = "MyPreference.Text")
+    @Property(bundleKey = "MyPreference.Text",
```
Comment (manstis) [inline uberfire-preferences/uberfire-preferences-backend/src/test/java/org/uberfire/preferences/backend/MyPreference.java]: Formatting

File: uberfire-preferences/uberfire-preferences-processors/src/test/resources/org/uberfire/ext/preferences/processors/MyPreference.java [hunk 4]
```diff
@@ -20,24 +20,32 @@
 import org.uberfire.preferences.shared.annotations.Property;
 import org.uberfire.preferences.shared.annotations.WorkbenchPreference;
 import org.uberfire.preferences.shared.bean.BasePreference;
+import org.uberfire.preferences.shared.impl.validation.NotEmptyValidator;
 
 @WorkbenchPreference(identifier = "MyPreference",
```
Comment (manstis) [inline uberfire-preferences/uberfire-preferences-processors/src/test/resources/org/uberfire/ext/preferences/processors/MyPreference.java]: Formatting.

File: uberfire-extensions/uberfire-wires/uberfire-wires-webapp/src/main/java/org/uberfire/ext/wires/shared/preferences/bean/MyPreference.java (hunk 1 above)
Comment (paulovmr) [inline uberfire-extensions/uberfire-wires/uberfire-wires-webapp/src/main/java/org/uberfire/ext/wires/shared/preferences/bean/MyPreference.java]: @manstis Actually I formatted the file, unfortunately this is how annotation members get formatted...

File: uberfire-preferences/uberfire-preferences-api/src/test/java/org/uberfire/preferences/shared/impl/validation/NotEmptyValidatorTest.java (hunk 2 above)
Comment (paulovmr) [inline uberfire-preferences/uberfire-preferences-api/src/test/java/org/uberfire/preferences/shared/impl/validation/NotEmptyValidatorTest.java]: @manstis Done.

File: uberfire-preferences/uberfire-preferences-backend/src/test/java/org/uberfire/preferences/backend/MyPreference.java (hunk 3 above)
Comment (paulovmr) [inline uberfire-preferences/uberfire-preferences-backend/src/test/java/org/uberfire/preferences/backend/MyPreference.java]: @manstis Actually I formatted the file, unfortunately this is how annotation members get formatted...

File: uberfire-preferences/uberfire-preferences-processors/src/test/resources/org/uberfire/ext/preferences/processors/MyPreference.java (hunk 4 above)
Comment (paulovmr) [inline uberfire-preferences/uberfire-preferences-processors/src/test/resources/org/uberfire/ext/preferences/processors/MyPreference.java]: @manstis Actually I formatted the file, unfortunately this is how annotation members get formatted...

File: uberfire-preferences/uberfire-preferences-api/src/main/java/org/uberfire/preferences/shared/bean/BasePreferencePortable.java [hunk 5]
```diff
@@ -74,6 +78,10 @@ void set(String property,
      */
     Map<String, PropertyFormType> getPropertiesTypes();
 
+    default Map<String, List<PropertyValidator>> getPropertiesValidators() {
```
Comment (ederign) [inline uberfire-preferences/uberfire-preferences-api/src/main/java/org/uberfire/preferences/shared/bean/BasePreferencePortable.java]: why this inner class?

Comment (paulovmr) [inline uberfire-preferences/uberfire-preferences-api/src/main/java/org/uberfire/preferences/shared/bean/BasePreferencePortable.java]: @ederign It's not a inner class, just a method with a default implementation. :)

Comment (ederign) [inline uberfire-preferences/uberfire-preferences-api/src/main/java/org/uberfire/preferences/shared/bean/BasePreferencePortable.java]: my bad.

**Pull request merged**